File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.py
# User function Template for python3

class Solution:
    def longestSubarray(self, m, k):
        x = {0:0}
        l = 0
        csum = 0
        for i in range(len(arr)):
            csum+=arr[i]
            if csum-k in x:
                l = max(l,i-x[csum-k]+1)
            if csum not in x:
                x[csum]=i+1
        return l

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    tc = int(input().strip())
    while tc > 0:
        arr = list(map(int, input().strip().split()))
        k = int(input().strip())
        ob = Solution()
        print(ob.longestSubarray(arr, k))
        tc -= 1
        print("~")
# } Driver Code Ends

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.java




// User function Template for Java


class Solution {
    public int longestSubarray(int[] arr, int K) {
        Map<Integer, Integer> prefixSumMap = new HashMap<>();
        int prefixSum = 0;
        int maxLength = 0;

        for (int i = 0; i < arr.length; i++) {
            prefixSum += arr[i];

            // अगर prefixSum खुद K है, तो subarray [0...i] valid है
            if (prefixSum == K) {
                maxLength = i + 1;
            }

            // Check if (prefixSum - K) exists in the map
            if (prefixSumMap.containsKey(prefixSum - K)) {
                int prevIndex = prefixSumMap.get(prefixSum - K);
                maxLength = Math.max(maxLength, i - prevIndex);
            }

            // Store the prefixSum if it is not already in the map
            if (!prefixSumMap.containsKey(prefixSum)) {
                prefixSumMap.put(prefixSum, i);
            }
        }

        return maxLength;
    }
}

